src/ThermalShallowWaterExplicit.jl

```diff
@@ -0,0 +1,219 @@
+  # L2MM       : L² mass matrix, ∫γγdΩ, ∀γ∈ L²(Ω)
+  # dt         : time step
+  # leap_frog  : do leap frog time integration for the first step (boolean)
+  # dΩ         : measure of the elements
+
+  n = get_normal_vector(model)
+  # explicit step for provisional velocity, uₚ
+  dt1 = dt
+  if leap_frog
+    dt1 = 2.0*dt
+  end
+
+  # 1.1: the mass flux
+  compute_mass_flux!(F,dΩ,V,RTMMchol,u₁*h₁)
+  # 1.2: the bernoulli function
+  compute_bernoulli_potential!(ϕ,dΩ,Q,L2MMchol,u₁⋅u₁,E₁,0.5)
+  # 1.3: materially advected quantities (potential vorticity and buoyancy)
+  compute_potential_vorticity!(q₁,H1h,H1hchol,dΩ,R,S,h₁,u₁,f,n)
+  compute_buoyancy!(e₁,dΩ,S,H1hchol,E₁)
+  # 1.4: solve for the provisional velocity
+  compute_temperature_gradient!(dT,dω,V,RTMMchol,0.5*h₁)
+  compute_velocity_tswe!(uₚ,dΩ,dω,V,RTMMchol,uₘ,q₁-τ*u₁⋅∇(q₁),e₁-τ*u₁⋅∇(e₁),F,ϕ,dT,n,dt1,dt1)
+  # 1.5: solve for the provisional depth
+  compute_depth!(hₚ,dΩ,dω,Q,L2MMchol,hₘ,F,dt1)
+  # 1.6: solve for the buoyancy weighted mass flux
+  compute_buoyancy_flux!(dT,dΩ,V,RTMMchol,e₁-τ*u₁⋅∇(e₁),F)
+  compute_depth!(Eₚ,dΩ,dω,Q,L2MMchol,Eₘ,dT,dt1)
+
+  # 2.1: the mass flux
+  compute_mass_flux!(F,dΩ,V,RTMMchol,u₁*(2.0*h₁ + hₚ)/6.0+uₚ*(h₁ + 2.0*hₚ)/6.0)
+  # 2.2: the bernoulli function
+  compute_bernoulli_potential!(ϕ,dΩ,Q,L2MMchol,(u₁⋅u₁ + u₁⋅uₚ + uₚ⋅uₚ)/3.0,0.5*(E₁ + Eₚ),0.5)
+  # 2.3: materially advected quantities (potential vorticity and buoyancy)
+  compute_potential_vorticity!(q₂,H1h,H1hchol,dΩ,R,S,hₚ,uₚ,f,n)
+  compute_buoyancy!(e₂,dΩ,S,H1hchol,Eₚ)
+  # 2.4: solve for the final velocity
+  compute_temperature_gradient!(dT,dω,V,RTMMchol,0.25*(h₁+hₚ))
+  compute_velocity_tswe!(u₂,dΩ,dω,V,RTMMchol,u₁,q₁-τ*u₁⋅∇(q₁)+q₂-τ*uₚ⋅∇(q₂),e₁-τ*u₁⋅∇(e₁)+e₂-τ*uₚ⋅∇(e₂),F,ϕ,dT,n,0.5*dt,dt)
+  # 2.5: solve for the final depth
+  compute_depth!(h₂,dΩ,dω,Q,L2MMchol,h₁,F,dt)
+  # 2.6: solve for the buoyancy weighted mass flux
+  compute_buoyancy_flux!(dT,dΩ,V,RTMMchol,0.5*(e₁-τ*u₁⋅∇(e₁)+e₂-τ*uₚ⋅∇(e₂)),F)
+  compute_depth!(E₂,dΩ,dω,Q,L2MMchol,E₁,dT,dt)
+end
+
+function thermal_shallow_water_explicit_time_stepper(model, order, degree,
+                        h₀, u₀, E₀, f₀,
+                        dt, τ, N;
+                        write_diagnostics=true,
+                        write_diagnostics_freq=1,
+                        dump_diagnostics_on_screen=true,
+                        write_solution=false,
+                        write_solution_freq=N/10,
+                        output_dir="tswe_ncells_$(num_cells(model))_order_$(order)_explicit")
+
+  # Forward integration of the shallow water equations
+  Ω = Triangulation(model)
+  dΩ = Measure(Ω, degree)
+  dω = Measure(Ω, degree, ReferenceDomain())
+
+  # Setup the trial and test spaces
+  R, S, U, V, P, Q = setup_mixed_spaces(model, order)
+
+  # assemble the mass matrices
+  H1MM, RTMM, L2MM, H1MMchol, RTMMchol, L2MMchol = setup_and_factorize_mass_matrices(dΩ, R, S, U, V, P, Q)
```

Comment:
I keep seeing the same pattern. `RTMM` is not needed outside the function, only inside for computing `RTMMchol`. Do we need to keep `RTMM` till the end of the simulation?